### PR TITLE
chore(storage): add more TaskStatus for metrics report

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- add more TaskStatus for metrics report
- seperate the fail and cancel report

## Checklist

- ~[ ] I have written necessary rustdoc comments~
- ~[ ] I have added necessary unit tests and integration tests~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- https://github.com/risingwavelabs/risingwave/issues/5303

### CODE REVIEW

File: proto/hummock.proto [hunk 1]
```diff
@@ -180,7 +180,14 @@ message CompactTask {
     PENDING = 0;
     SUCCESS = 1;
     FAILED = 2;
```
Comment (Gun9niR) [inline proto/hummock.proto:182]: Do we still need `FAILED`?

File: src/meta/src/hummock/manager/mod.rs [hunk 2]
```diff
@@ -716,13 +719,17 @@ where
...
+        compact_task: &mut CompactTask,
+        task_status: TaskStatus,
+    ) -> Result<bool> {
+        compact_task.set_task_status(task_status);
         self.cancel_compact_task_impl(compact_task).await
     }
 
     pub async fn cancel_compact_task_impl(&self, compact_task: &CompactTask) -> Result<bool> {
-        assert_eq!(compact_task.task_status(), TaskStatus::Canceled);
+        // assert_eq!(compact_task.task_status(), TaskStatus::Canceled);
```
Comment (Gun9niR) [inline src/meta/src/hummock/manager/mod.rs:732]: Maybe we can remove this line.

File: src/meta/src/hummock/manager/mod.rs [hunk 3]
```diff
@@ -716,13 +719,17 @@ where
     }
 
     /// Cancels a compaction task no matter it's assigned or unassigned.
-    pub async fn cancel_compact_task(&self, compact_task: &mut CompactTask) -> Result<bool> {
-        compact_task.set_task_status(TaskStatus::Canceled);
+    pub async fn cancel_compact_task(
+        &self,
+        compact_task: &mut CompactTask,
+        task_status: TaskStatus,
```
Comment (Gun9niR) [inline src/meta/src/hummock/manager/mod.rs:725]: Maybe add an assertion to make sure `task_status` is one of those CANCELED statuses?

**Review by Gun9niR — APPROVED** [review]
REST LGTM.

**Review by zwang28 — APPROVED** [review]

File: src/meta/src/hummock/manager/mod.rs (hunk 3 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/mod.rs:757]: Thanks, add assert in `cancel_compact_task_impl `

**Comment by Li0k** [comment]
image

**Comment by Li0k** [comment]
image

File: src/meta/src/hummock/compaction_scheduler.rs [hunk 4]
```diff
@@ -159,25 +160,33 @@ where
         let schedule_status = self
             .pick_and_assign_impl(compaction_group, request_channel)
             .await;
-        let cancel_task = match &schedule_status {
-            ScheduleStatus::Ok => None,
-            ScheduleStatus::NoTask | ScheduleStatus::PickFailure => None,
-            ScheduleStatus::NoAvailableCompactor(task)
-            | ScheduleStatus::AssignFailure(task)
-            | ScheduleStatus::SendFailure(task) => Some(task.clone()),
+        let (cancel_task, cancel_task_state) = match &schedule_status {
```
Comment (hzxa21) [inline src/meta/src/hummock/compaction_scheduler.rs:163]: nits: we can return a Option<(task, status)> instead of (Option<task>, Option<status>)

**Review by hzxa21 — APPROVED** [review]
Rest LGTM

File: src/meta/src/hummock/compaction_scheduler.rs (hunk 4 above)
Comment (Li0k) [reply src/meta/src/hummock/compaction_scheduler.rs]: Fix it

**Pull request merged**